File: Java/ValidParentheses.java
public class Solution {
    public boolean isValid(String s) {
        if (s.length() % 2 != 0) return false;
    
        HashMap<Character,Character> matchingParensMap = new HashMap<Character, Character>();
        matchingParensMap.put(')' , '(');
        matchingParensMap.put('}' , '{');
        matchingParensMap.put(']' , '[');  
    
        Stack<Character> stackOfParen = new Stack<Character>();
        
        for (char ch : s) {
            if(matchingParensMap.containsValue( ch ) {
                stackOfParen.push( ch );
            } else if( stackOfParen.empty() || matchingParensMap.get( ch ) != stackOfParen.pop() ) {
                return false;
            }
        }
        return stackOfParen.empty();
    }
}

File: C++/ValidParentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        std::set<char>rightPerens{']' , '}' , ')'};
        std::map<char,char>leftToRightPerens {
            { '[' , ']' },
            { '{' , '}' },
            { '(' , ')' } };
        std::stack<char> perens{};
        
        int position = s.size();
        while(position){
            char temp = s.at(--position);
            
            if(rightPerens.end() != rightPerens.find(temp))
                    perens.push(temp);
            else{
                if(perens.size() == 0 || perens.top() != leftToRightPerens[temp])
                    return false;
                perens.pop();
            }
        }
        return perens.size() == 0;
    }
};
